File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.js
/**
 * @param {number} n
 * @param {number[][]} reservedSeats
 * @return {number}
 */
var maxNumberOfFamilies = function(n, reservedSeats) {
    const rows = new Map();

    for (const [row, col] of reservedSeats) {
        if (col >= 2 && col <= 11) {
            const mask = rows.get(row) || 0;
            rows.set(row, mask | (1 << (col - 2)));
        }
    }

    let answer = (n - rows.size) * 2;

    for (const mask of rows.values()) {
        const left = (mask & 0b00001111) === 0;
        const middle = (mask & 0b00111100) === 0;
        const right = (mask & 0b11110000) === 0;

        if (left && right) {
            answer += 2;
        }
        else if (left || middle || right) {
            answer += 1;
        }
    }

    return answer;
};

File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.py
from typing import List

class Solution:
    def maxNumberOfFamilies(
        self,
        n: int,
        reservedSeats: List[List[int]]
    ) -> int:

        rows = {}

        for row, col in reservedSeats:
            if 2 <= col <= 11:
                rows[row] = rows.get(row, 0) | (1 << (col - 2))

        # Every completely empty row can fit 2 families.
        answer = (n - len(rows)) * 2

        for mask in rows.values():

            left = (mask & 0b00001111) == 0
            middle = (mask & 0b00111100) == 0
            right = (mask & 0b11110000) == 0

            if left and right:
                answer += 2
            elif left or middle or right:
                answer += 1

        return answer